Shared extension for the Hibernate Search Elasticsearch backend

yrodiere opened the pull request:
To reduce code duplication.

~~Opening as draft, because:~~

* ~~It's based on #39413 and #39415, which must be merged first.~~ => Done
* ~~It requires #35439 to get fixed first, and that will take a while; see https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Reusing.20ConfigMapping.20from.20other.20JARS~~ => Done

~~Based on #43943, which must be merged first.~~ => Done

yrodiere committed:
Move Hibernate Search Elasticsearch backend to a dedicated extension

yrodiere committed:
Fix typos in documentation of configuration properties

yrodiere commented:
Rebased on main. Ran a few tests locally, and it seems to be working fine... we'll need to check the generated documentation for configuration properties though.

marko-bekhta commented on extensions/hibernate-search-backend-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/backend/elasticsearch/runtime/HibernateSearchBackendElasticsearchRuntimeConfig.java:
```diff
@@ -0,0 +1,351 @@
+         * which may lead to higher indexing throughput,
+         * but incurs a risk of overloading Elasticsearch,
+         * i.e. of overflowing its HTTP request buffers and tripping
+         * https://www.elastic.co/guide/en/elasticsearch/reference/7.9/circuit-breaker.html[circuit breakers],
```
I wonder... (not for this PR)
since this is an asciidoc comment, should we do something similar as with `hibernate-search-docs-url` or maybe just point to `current` (https://www.elastic.co/guide/en/elasticsearch/reference/current/circuit-breaker.html)
(I thought about it as 7.9 is slightly EOLed 🙈 😃 )

yrodiere replied:
I agree it's a good idea, but would likely require moving Maven properties for the Elasticsearch version to the root POM: for a different PR indeed.

marko-bekhta commented on extensions/hibernate-search-backend-elasticsearch/deployment/src/main/java/io/quarkus/hibernate/search/backend/elasticsearch/deployment/HibernateSearchBackendElasticsearchProcessor.java:
```diff
@@ -0,0 +1,153 @@
+class HibernateSearchBackendElasticsearchProcessor {
+
+    @BuildStep
+    void registerBeans(List<HibernateSearchBackendElasticsearchEnabledBuildItem> searchEnabledPUs,
```
```suggestion
    void registerBeans(List<HibernateSearchBackendElasticsearchEnabledBuildItem> elasticsearchEnabled,
```
a bit of nitpicking 🙈 feel free to ignore 😄

marko-bekhta commented on extensions/hibernate-search-orm-elasticsearch/deployment/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/deployment/HibernateSearchElasticsearchProcessor.java:
```diff
@@ -161,8 +148,19 @@ private void buildForPersistenceUnit(HibernateSearchElasticsearchRecorder record
         configuredPersistenceUnits
-                .produce(new HibernateSearchElasticsearchPersistenceUnitConfiguredBuildItem(persistenceUnitName, puConfig,
-                        backendNamesForIndexedEntities, backendAndIndexNamesForSearchExtensions));
+                .produce(new HibernateSearchElasticsearchPersistenceUnitConfiguredBuildItem(
```
I've looked at the failing test... maybe we should fail early (`puConfig== null` ?), if there's no config for a PU, since we know that we need at least some configuration to be present.

yrodiere replied:
For now I'll just add a null-check, but yes it would be a good idea to move the version check to build time.

marko-bekhta reviewed (commented):
This is a nice improvement! 😃 
There are a couple of failing tests because of the missing pu-config, I'll let you decide how you want to address those 😃. Other than that, it looks good!

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/runtime/HibernateSearchConfigUtil.java:
```diff
@@ -8,6 +8,10 @@
+/**
+ * @deprecated Use {@link io.quarkus.hibernate.search.backend.elasticsearch.runtime.HibernateSearchConfigUtil} instead.
+ */
+@Deprecated
```
I don't think we approach them at all :]
Personally I just use Git history for this kind of things.

yrodiere commented:
Thanks for your comments @marko-bekhta , I think I addressed most of them.

Now, the failure to generate documentation is actually more serious than it seemed... I started working on a potential solution, but in essence we're lacking some features in Quarkus docs generation. I'll try to discuss this with Guillaume, see #43808 for details (or don't, and save your time for something more interesting xD).

yrodiere commented:
@gsmet I rebased my branch (this PR) on main and removed any reliance on my work in ##43808. You should be able to test your `xref:#` link rewriting by rebasing it again on your work.
FWIW, relevant configuration options whose documention include links to the guide:

* `io.quarkus.hibernate.search.backend.elasticsearch.runtime.HibernateSearchBackendElasticsearchRuntimeConfig.LayoutConfig#strategy`:
  * `quarkus.hibernate-search-orm.elasticsearch.layout.strategy`
  * `quarkus.hibernate-search-standalone.elasticsearch.layout.strategy`
  * Two links: "bean reference", "thisection".
* `io.quarkus.hibernate.search.backend.elasticsearch.runtime.HibernateSearchBackendElasticsearchBuildTimeConfig.AnalysisConfig#configurer`:
  * `quarkus.hibernate-search-orm.elasticsearch.analysis.configurer`
  * `quarkus.hibernate-search-standalone.elasticsearch.analysis.configurer`
  * Three links: "bean reference", "analysis-configurer", "this section".

Thanks for having a look!

yrodiere commented:
#43943 was merged, I rebased on main, so I think we're all set.

@marko-bekhta I think I addressed your comments, do you mind approving please?

marko-bekhta reviewed (approved):
🎉

gsmet commented:
I'll have a look later today.

gsmet commented on extensions/hibernate-search-backend-elasticsearch/deployment/pom.xml:
```diff
@@ -0,0 +1,52 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-hibernate-search-backend-elasticsearch-deployment</artifactId>
+    <name>Quarkus - Hibernate Search - Elasticsearch - Deployment</name>
```
I have only one comment for this PR: for common/internal extensions that are not to be consumed, we usually have a `-common` or `-internal` suffix to the artifactId.

I actually use it in the config doc generation (using unlisted is unfortunately not the semantic I need). I think it would be worth adding the `-common` suffix here.

yrodiere replied:
Thanks for the review.

> I think it would be worth adding the -common suffix here.

I think it would be weird since there is no other artifact starting with `quarkus-hibernate-search-backend-elasticsearch`... So there's no obvious answer to the question "common to what?".

But if the tooling requires it, so be it.
Users aren't supposed to reference this artifact directly anyway, so the weirdness shouldn't reach them.

gsmet reviewed (commented):
Sorry for the very late review. I added only one (annoying, sorry...) comment. I think it would be worth taking it into account even if I know you won't be very excited about it :).

yrodiere committed:
Rename quarkus-hibernate-search-backend-elasticsearch to quarkus-hibernate-search-backend-elasticsearch-common

yrodiere reviewed (commented):
I renamed the extension to `quarkus-hibernate-search-backend-elasticsearch-common` as requested.

gsmet commented:
I somehow forgot about this one...

yrodiere commented:
No harm done: so did I... :)
